src/hotspot/share/gc/shared/stringdedup/stringDedup.cpp

```diff
@@ -21,61 +21,178 @@
-  StringDedupTable::unlink_or_oops_do(unlink, worker_id);
+void StringDedup::notify_intern(oop java_string) {
+  assert(is_enabled(), "precondition");
+  // A String that is interned in the StringTable must not later have its
+  // underlying byte array changed, so mark it as not deduplicatable.  But we
+  // can still add the byte array to the dedup table for sharing, so add the
+  // string to the pending requests.  Triggering request processing is left
+  // to the next GC.
+  {
+    MutexLocker ml(StringDedupIntern_lock, Mutex::_no_safepoint_check_flag);
+    java_lang_String::set_deduplication_forbidden(java_string);
+  }
+  StorageUse* requests = Processor::storage_for_requests();
+  oop* ref = requests->storage()->allocate();
+  if (ref != nullptr) {
+    NativeAccess<ON_PHANTOM_OOP_REF>::oop_store(ref, java_string);
+    log_trace(stringdedup)("StringDedup::deduplicate");
+  }
+  requests->relinquish();
 }
 
-void StringDedup::threads_do(ThreadClosure* tc) {
-  assert(is_enabled(), "String deduplication not enabled");
-  tc->do_thread(StringDedupThread::thread());
+StringDedup::Requests::Requests() :
+  _storage_for_requests(nullptr), _buffer(nullptr), _index(0), _active(true)
+{}
+
+StringDedup::Requests::~Requests() {
+  flush();
 }
 
-void StringDedup::verify() {
-  assert(is_enabled(), "String deduplication not enabled");
-  StringDedupQueue::verify();
-  StringDedupTable::verify();
+void StringDedup::Requests::add(oop java_string) {
+  assert(is_enabled(), "StringDedup not enabled");
+  const size_t buffer_size = OopStorage::bulk_allocate_limit;
+  if (_index == 0) {
+    if (!_active) return;
+    // Lazy initialization of the requests object.  It can be common for
+    // many of the marking threads to not encounter any candidates.
+    if (_buffer == nullptr) {
+      // Lazily allocate a buffer to hold pre-allocated storage entries.
+      _buffer = NEW_C_HEAP_ARRAY_RETURN_NULL(oop*, buffer_size, mtStringDedup);
+      if (_buffer == nullptr) {
+        log_debug(stringdedup)("request failed to allocate buffer");
+        _active = false;
+        return;
+      }
+      // Lazily obtain the storage object to use for requests.
+      assert(_storage_for_requests == nullptr, "invariant");
+      _storage_for_requests = Processor::storage_for_requests();
+    }
+    assert(_storage_for_requests != nullptr, "invariant");
+    // Bulk pre-allocate some storage entries to satisfy this and future
+    // requests.  This amortizes the cost of allocating entries over
+    // multiple requests, and reduces contention on the storage object.
+    _index = _storage_for_requests->storage()->allocate(_buffer, buffer_size);
+    if (_index == 0) {
+      log_debug(stringdedup)("request failed to allocate oopstorage entries");
+      flush();
+      _active = false;
+      return;
+    }
+  }
+  // Store the string in the next pre-allocated storage entry.
+  oop* ref = _buffer[--_index];
+  _buffer[_index] = nullptr;
```

Comment:
It's not really needed to clear the slot with null, right?

Reply:
You are right; there used to be some asserts, but no longer. Removed.